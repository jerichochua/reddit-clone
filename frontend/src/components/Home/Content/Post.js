import React from 'react';

export default function Post(props) {
  return (
    <li>
      <div className='post-container'>
        <div className='votes'>
          <button className='upvote-btn'></button>
          <span>{props.votes}</span>
          <button className='downvote-btn'></button>
        </div>
        <div className='post-item'>
          <p className='title'>{props.title}</p>
          <p className='caption'>
            by {props.author} {props.timestamp}
          </p>
          <p className='links'>{props.comments} comments</p>
        </div>
      </div>
    </li>
  );
}
