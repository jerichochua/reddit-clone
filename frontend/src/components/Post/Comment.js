import React from 'react';
import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

dayjs.extend(relativeTime);

const Comment = ({ comment, username, onDelete }) => {
  const isAuthor = username === comment.author;

  return (
    <li className='comments-list-item'>
      <div className='comment-container'>
        <div className='comment-header'>
          <span className='comment-header-author'>{comment.author}</span>
          <span className='comment-header-timestamp'>
            {dayjs(comment.created_at).fromNow()}
          </span>
          {isAuthor && (
            <button
              value={comment.id}
              className='delete-button'
              onClick={onDelete}
            >
              Delete
            </button>
          )}
        </div>
        <div className='comment-body'>
          <p>{comment.content}</p>
        </div>
      </div>
    </li>
  );
};

export default Comment;
