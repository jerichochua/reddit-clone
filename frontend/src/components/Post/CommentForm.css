.comment-form-container {
  margin-top: -1px;
  border: 1px solid rgb(235, 237, 240);
  transition: border 0.1s ease-in-out;
}

.comment-form-container:hover {
  border: 1px solid rgb(0, 149, 246);
}

.comment-form-container:focus-within {
  border: 1px solid rgb(0, 149, 246);
}

.comment-form {
  display: flex;
  flex-direction: column;
}

.comment-form-textarea {
  width: 100%;
  padding: 8px;
  background-color: rgb(250, 250, 250);
  font-size: 15px;
  resize: none;
  outline: none;
  border: none;
  border-bottom: 1px solid rgb(235, 237, 240);
}

.comment-form-button {
  padding: 8px 24px;
  margin: 4px;
  font-size: 12px;
  font-weight: 700;
  text-transform: uppercase;
  border: none;
  border-radius: 4px;
  background-color: rgb(0, 149, 246);
  color: white;
  cursor: pointer;
  align-self: flex-end;
}
