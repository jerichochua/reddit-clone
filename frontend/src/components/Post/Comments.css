.comments-list {
  margin-top: 16px;
  list-style: none;
}

.comments-list-item {
  margin-bottom: 8px;
}

.comment-container {
  border: 1px solid rgb(235, 237, 240);
}

.comment-header {
  display: flex;
  padding: 8px;
  border-bottom: 1px solid rgb(235, 237, 240);
  background-color: rgb(244, 246, 248);
}

.comment-header-author {
  font-weight: 500;
  font-size: 13px;
}

.comment-header-timestamp {
  margin-left: 4px;
  margin-right: auto;
  color: rgb(129, 142, 153);
  font-size: 13px;
}

.comment-body {
  padding: 12px;
  font-size: 15px;
  white-space: pre-line;
  overflow-wrap: break-word;
}

.comment-form-footer {
  display: flex;
  justify-content: flex-end;
}

.comment-footer-message {
  align-self: center;
  margin-right: 12px;
  color: rgb(129, 142, 153);
  font-size: 13px;
  text-transform: lowercase;
}

.comment-edit-textarea {
  width: 100%;
  resize: none;
  font-size: 15px;
  font-family: inherit;
  padding: 15px;
  border: 1px solid rgb(0, 149, 246);
  outline: none;
}

.comment-edit-footer {
  display: flex;
  justify-content: flex-end;
}

