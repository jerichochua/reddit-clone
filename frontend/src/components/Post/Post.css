.post {
  display: flex;
  flex-direction: column;
  padding: 0 10vw;
}

@media (max-width: 1024px) {
  .post {
    padding: 0 5vw;
  }
}

@media (max-width: 768px) {
  .post {
    padding: 0;
  }
}

.post-details-container {
  display: flex;
  border: 1px solid rgb(235, 237, 240);
  flex: 1 1;
}

.post-content-container {
  display: flex;
  flex-direction: column;
  flex: 1 1;
  padding: 8px;
  border-left: 1px solid rgb(235, 237, 240);
}

.post-title {
  display: flex;
}

.post-title span {
  font-weight: 500;
}

.post-content {
  margin: 8px -8px;
  padding: 8px;
  border-top: 1px solid rgb(235, 237, 240);
  border-bottom: 1px solid rgb(235, 237, 240);
  white-space: pre-line;
}

.post-details {
  display: flex;
  font-size: 13px;
  color: rgb(129, 142, 153);
}

.delete-button {
  margin-left: auto;
  font-size: 13px;
  font-weight: 700;
  text-transform: uppercase;
  background-color: transparent;
  color: red;
  border: none;
  cursor: pointer;
}
