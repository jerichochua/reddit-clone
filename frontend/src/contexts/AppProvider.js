import React, { createContext, useContext, useReducer } from 'react';
import jwt_decode from 'jwt-decode';

const AppContext = createContext();
AppContext.displayName = 'AppContext';

const token = localStorage.getItem('token');
const userId = token ? jwt_decode(token).id : null;
const username = token ? jwt_decode(token).username : null;

const initialState = {
  userId,
  username,
  token,
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_USER':
      return {
        ...state,
        userId: action.payload.userId,
        username: action.payload.username,
      };
    case 'SET_TOKEN':
      return { ...state, token: action.payload };
    case 'LOGOUT':
      return { ...state, userId: null, username: null, token: null };
    default:
      return state;
  }
};

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  return useContext(AppContext);
};
