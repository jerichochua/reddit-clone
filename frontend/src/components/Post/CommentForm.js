import React, { useState } from 'react';
import { post } from '../../util/api';
import { FormButton } from '../Form';
import { useAppContext } from '../../contexts/AppProvider';
import './CommentForm.css';

const CommentForm = ({ postId }) => {
  const { state } = useAppContext();
  const [message, setMessage] = useState('');

  const onSubmitForm = async (e) => {
    e.preventDefault();
    const body = { content: e.target.content.value };
    try {
      const response = await post(`posts/${postId}`, body, state.token);
      if (response) {
        setMessage('comment added');
        e.target.content.value = '';
      } else {
        setMessage('error adding comment');
        console.error(response);
      }
    } catch (error) {
      setMessage(`error adding comment: ${error.message}`);
    }
  };

  return (
    <div className='comment-form-container'>
      <form className='comment-form' onSubmit={onSubmitForm}>
        <textarea
          name='content'
          className='comment-form-textarea'
          placeholder='Add a comment...'
          rows='2'
          required
        ></textarea>
        <div className='comment-form-footer'>
          {message && <span className='comment-footer-message'>{message}</span>}
          <FormButton label='Comment' type='submit' />
        </div>
      </form>
    </div>
  );
};

export default CommentForm;
