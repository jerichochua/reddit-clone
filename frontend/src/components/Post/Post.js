import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { get, delete_request } from '../../util/api';
import { useAppContext } from '../../contexts/AppProvider';
import CommentForm from './CommentForm';
import CommentsList from './CommentsList';
import Empty from '../Empty/Empty';
import Toast from '../Toast/Toast';
import PostDetails from './PostDetails';
import './Post.css';

const Post = () => {
  const { state } = useAppContext();
  const { id } = useParams();
  const [post, setPost] = useState({});
  const [error, setError] = useState(false);
  const [deleteSuccess, setDeleteSuccess] = useState(false);
  const [deleteError, setDeleteError] = useState(false);
  const [editSuccess, setEditSuccess] = useState(false);
  const [editError, setEditError] = useState(false);
  const [message, setMessage] = useState('');

  const onDeletePost = async () => {
    try {
      await delete_request(`posts/${id}`, state.token);
      setDeleteError(false);
      setDeleteSuccess(true);
      setMessage('Post deleted successfully. Redirecting to home page...');
      setTimeout(() => {
        window.location.href = '/';
      }, 2000);
    } catch (err) {
      setDeleteError(true);
      setDeleteSuccess(false);
      setMessage('Unable to delete post');
    }
  };

  useEffect(() => {
    const fetchPost = async () => {
      try {
        const response = await get(`posts/${id}`);
        setPost(response);
      } catch (err) {
        setError(true);
      }
    };
    fetchPost();
  }, [id]);

  return (
    <main className='post'>
      {error ? (
        <Empty message='Post not found' />
      ) : (
        <>
          <PostDetails post={post} onDelete={onDeletePost} />
          {state.token && <CommentForm postId={id} />}
          <CommentsList
            state={state}
            postId={id}
            setMessage={setMessage}
            setDeleteError={setDeleteError}
            setDeleteSuccess={setDeleteSuccess}
            setEditError={setEditError}
            setEditSuccess={setEditSuccess}
          />
          {deleteError && (
            <Toast type='error' message={message} show={deleteError} />
          )}
          {deleteSuccess && (
            <Toast type='info' message={message} show={deleteSuccess} />
          )}
          {editError && (
            <Toast type='error' message={message} show={editError} />
          )}
          {editSuccess && (
            <Toast type='info' message={message} show={editSuccess} />
          )}
        </>
      )}
    </main>
  );
};

export default Post;
