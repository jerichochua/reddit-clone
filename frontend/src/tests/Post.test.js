import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { AppProvider } from '../contexts/AppProvider';
import Post from '../components/Post/Post';

it('renders correctly', () => {
  const component = renderer.create(
    <MemoryRouter>
      <AppProvider>
        <Post />
      </AppProvider>
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
