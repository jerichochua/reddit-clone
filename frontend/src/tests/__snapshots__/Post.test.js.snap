// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="post"
>
  <div
    className="post-details-container"
  >
    <div
      className="votes-container"
    >
      <button
        className="upvote-btn"
        onClick={[Function]}
      />
      <span />
      <button
        className="downvote-btn"
        onClick={[Function]}
      />
    </div>
    <div
      className="post-content-container"
    >
      <div
        className="post-title"
      >
        <span />
      </div>
      <div
        className="post-content"
      />
      <div
        className="post-details"
      >
        <p>
           comment
          s
           by
           
           
          a few seconds ago
        </p>
      </div>
    </div>
  </div>
  <ul
    className="comments-list"
  />
</main>
`;
